File: imogi_finance/imogi_finance/report/budget_control_dashboard/budget_control_dashboard.js
// Copyright (c) 2026, PT. Inovasi Terbaik Bangsa and contributors
// For license information, please see license.txt

frappe.query_reports["Budget Control Dashboard"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
			"reqd": 1
		},
		{
			"fieldname": "fiscal_year",
			"label": __("Fiscal Year"),
			"fieldtype": "Link",
			"options": "Fiscal Year",
			"reqd": 1,
			"get_query": function() {
				return {
					"filters": {
						"disabled": 0
					}
				}
			}
		},
		{
			"fieldname": "cost_center",
			"label": __("Cost Center"),
			"fieldtype": "Link",
			"options": "Cost Center",
			"reqd": 0
		},
		{
			"fieldname": "account",
			"label": __("Account"),
			"fieldtype": "Link",
			"options": "Account",
			"reqd": 0,
			"get_query": function() {
				return {
					"filters": {
						"is_group": 0,
						"root_type": "Expense"
					}
				}
			}
		},
		{
			"fieldname": "project",
			"label": __("Project"),
			"fieldtype": "Link",
			"options": "Project",
			"reqd": 0
		},
		{
			"fieldname": "branch",
			"label": __("Branch"),
			"fieldtype": "Link",
			"options": "Branch",
			"reqd": 0
		},
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"reqd": 0
		},
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"reqd": 0
		},
		{
			"fieldname": "hide_zero",
			"label": __("Hide Zero Balances"),
			"fieldtype": "Check",
			"default": 0
		}
	],
	
	"onload": function(report) {
		// Add help button with detailed guide
		report.page.add_inner_button(__('📖 Panduan Dashboard'), function() {
			frappe.msgprint({
				title: __('Cara Membaca Budget Control Dashboard'),
				indicator: 'blue',
				message: `
					<div style="font-size: 13px; line-height: 1.7;">
						<h4 style="margin-top: 0; color: #2e7d32;">📊 Formula Perhitungan</h4>
						<div style="background: #f5f5f5; padding: 12px; border-radius: 4px; margin-bottom: 15px; font-family: monospace;">
							<strong>Net Reserved</strong> = Reservation - Consumption + Reversal<br>
							<strong>Committed</strong> = Actual (GL) + Net Reserved<br>
							<strong>Available</strong> = Allocated - Committed
						</div>
						
						<h4 style="color: #1976d2;">💡 Arti Kolom Budget Control Entries</h4>
						<ul style="margin-bottom: 15px;">
							<li><strong style="color: #ff9800;">Reservation (+)</strong>: Budget di-lock saat Expense Request approved<br>
							    <small>Entry OUT - Mengurangi available</small></li>
							<li><strong style="color: #2196f3;">Consumption (-)</strong>: Budget dikonsumsi saat Purchase Invoice submit<br>
							    <small>Entry IN - Mengurangi reservation (bukan available!)</small></li>
							<li><strong style="color: #4caf50;">Reversal (+)</strong>: Budget dikembalikan saat Purchase Invoice cancel<br>
							    <small>Entry OUT - Menambah kembali reservation</small></li>
						</ul>
						
						<h4 style="color: #e91e63;">🎯 Contoh Skenario</h4>
						<div style="background: #fff3e0; padding: 12px; border-radius: 4px; margin-bottom: 15px;">
							<strong>Allocated: 100jt</strong><br><br>
							
							<strong>1. ER Submit (30jt)</strong><br>
							• Reservation: +30jt<br>
							• Net Reserved: 30jt<br>
							• Available: 70jt<br><br>
							
							<strong>2. PI Submit (30jt)</strong><br>
							• Consumption: -30jt (mengurangi reservation!)<br>
							• Actual: 30jt (dari GL Entry)<br>
							• Net Reserved: 30jt - 30jt = 0jt<br>
							• Available: 100jt - 30jt - 0jt = 70jt ✅ (tetap!)<br><br>
							
							<strong>3. PI Cancel</strong><br>
							• Reversal: +30jt<br>
							• Actual: 0jt (GL reversed)<br>
							• Net Reserved: 30jt - 30jt + 30jt = 30jt<br>
							• Available: 70jt ✅ (kembali!)
						</div>
						
						<h4 style="color: #9c27b0;">📈 Status Budget</h4>
						<ul style="list-style: none; padding-left: 0;">
							<li><span class="indicator-pill grey">Unused</span> - Budget belum dipakai</li>
							<li><span class="indicator-pill blue">In Use</span> - Normal, <75% committed</li>
							<li><span class="indicator-pill yellow">Warning</span> - Perhatian, 75-90% committed</li>
							<li><span class="indicator-pill orange">Critical</span> - Kritis, >90% committed</li>
							<li><span class="indicator-pill red">Over Budget</span> - Sudah melebihi alokasi!</li>
						</ul>
						
						<div style="background: #e3f2fd; padding: 12px; border-radius: 4px; margin-top: 15px;">
							<strong>💡 Tips Monitoring:</strong>
							<ul>
								<li>Available > 20% = Budget aman ✅</li>
								<li>Net Reserved > 50% = Banyak ER pending, percepat proses PI 🔄</li>
								<li>Committed > 90% = Stop approve ER baru ⛔</li>
							</ul>
						</div>
					</div>
				`
			});
		}, __('Help'));
	},
	
	"formatter": function(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		
		// Skip formatting for total row or if data is undefined
		if (!data || data.isTotal || row.isTotal) {
			return value;
		}
		
		// Color code Budget Control Entry types
		// Reservation: Orange (positive = mengurangi available)
		if (column.fieldname == "reservation" && data.reservation !== undefined && data.reservation !== 0) {
			const color = data.reservation > 0 ? "#ff9800" : "#d9534f";
			value = `<span style="color: ${color}; font-weight: bold;">${frappe.format(data.reservation, {fieldtype: 'Currency'})}</span>`;
		}
		
		// Consumption: Blue when negative (consuming reservation), Red when positive
		if (column.fieldname == "consumption" && data.consumption !== undefined && data.consumption !== 0) {
			const color = data.consumption < 0 ? "#2196f3" : "#d9534f";
			value = `<span style="color: ${color}; font-weight: bold;">${frappe.format(data.consumption, {fieldtype: 'Currency'})}</span>`;
		}
		
		// Reversal: Green (positive = restoring reservation)
		if (column.fieldname == "reversal" && data.reversal !== undefined && data.reversal !== 0) {
			const color = data.reversal > 0 ? "#4caf50" : "#d9534f";
			value = `<span style="color: ${color}; font-weight: bold;">${frappe.format(data.reversal, {fieldtype: 'Currency'})}</span>`;
		}
		
		if (column.fieldname == "reclass" && data.reclass !== undefined && data.reclass !== 0) {
			value = `<span style="color: #5bc0de;">${frappe.format(data.reclass, {fieldtype: 'Currency'})}</span>`;
		}
		
		if (column.fieldname == "supplement" && data.supplement !== undefined && data.supplement > 0) {
			value = `<span style="color: #5cb85c; font-weight: bold;">${frappe.format(data.supplement, {fieldtype: 'Currency'})}</span>`;
		}
		
		if (column.fieldname == "net_reserved" && data.net_reserved !== undefined) {
			if (data.net_reserved < 0) {
				value = `<span style="color: #d9534f; font-weight: bold;">${frappe.format(data.net_reserved, {fieldtype: 'Currency'})}</span>`;
			} else if (data.net_reserved > 0) {
				value = `<span style="color: #f0ad4e; font-weight: bold;">${frappe.format(data.net_reserved, {fieldtype: 'Currency'})}</span>`;
			}
		}
		
		// Color code status
		if (column.fieldname == "status" && data.status) {
			if (data.status == "Over Budget" || data.status == "Over Committed") {
				value = `<span class="indicator-pill red filterable" data-filter="status,=,${data.status}">${data.status}</span>`;
			} else if (data.status == "Critical") {
				value = `<span class="indicator-pill orange filterable" data-filter="status,=,Critical">${data.status}</span>`;
			} else if (data.status == "Warning") {
				value = `<span class="indicator-pill yellow filterable" data-filter="status,=,Warning">${data.status}</span>`;
			} else if (data.status == "In Use") {
				value = `<span class="indicator-pill blue filterable" data-filter="status,=,In Use">${data.status}</span>`;
			} else if (data.status == "Unused") {
				value = `<span class="indicator-pill grey filterable" data-filter="status,=,Unused">${data.status}</span>`;
			} else if (data.status == "Fully Used") {
				value = `<span class="indicator-pill darkgrey filterable" data-filter="status,=,Fully Used">${data.status}</span>`;
			} else {
				value = `<span class="indicator-pill grey">${data.status}</span>`;
			}
		}
		
		// Color code available
		if (column.fieldname == "available" && data.available !== undefined) {
			if (data.available < 0) {
				value = `<span style="color: #d9534f; font-weight: bold;">${frappe.format(data.available, {fieldtype: 'Currency'})}</span>`;
			} else if (data.available == 0) {
				value = `<span style="color: #f0ad4e; font-weight: bold;">${frappe.format(data.available, {fieldtype: 'Currency'})}</span>`;
			} else {
				value = `<span style="color: #5cb85c;">${frappe.format(data.available, {fieldtype: 'Currency'})}</span>`;
			}
		}
		
		// Color code committed percentage
		if (column.fieldname == "committed_pct") {
			let pct_value = data[column.fieldname];
			if (pct_value !== undefined && pct_value !== null && !isNaN(pct_value)) {
				if (pct_value >= 100) {
					value = `<span style="color: #d9534f; font-weight: bold;">${pct_value.toFixed(1)}%</span>`;
				} else if (pct_value > 90) {
					value = `<span style="color: #f0ad4e; font-weight: bold;">${pct_value.toFixed(1)}%</span>`;
				} else if (pct_value > 75) {
					value = `<span style="color: #ffc107;">${pct_value.toFixed(1)}%</span>`;
				} else {
					value = `<span style="color: #5cb85c;">${pct_value.toFixed(1)}%</span>`;
				}
			}
		}
		
		return value;
	},
	
	"onload": function(report) {
		// Set default filters if not already set
		if (!report.get_filter_value('company')) {
			let default_company = frappe.defaults.get_user_default("Company");
			if (default_company) {
				report.set_filter_value('company', default_company);
			}
		}
		
		if (!report.get_filter_value('fiscal_year')) {
			// Get current fiscal year
			frappe.call({
				method: 'erpnext.accounts.utils.get_fiscal_year',
				args: {
					date: frappe.datetime.get_today(),
					company: report.get_filter_value('company') || frappe.defaults.get_user_default("Company")
				},
				callback: function(r) {
					if (r.message && r.message[0]) {
						report.set_filter_value('fiscal_year', r.message[0]);
					}
				}
			});
		}
		
		// Add custom button to view details
		report.page.add_inner_button(__("View Budget Control Entries"), function() {
			frappe.set_route("List", "Budget Control Entry");
		});
		
		// Add refresh button
		report.page.add_inner_button(__("Refresh"), function() {
			report.refresh();
		});
	}
};

File: imogi_finance/imogi_finance/imogi_finance/report/budget_control_dashboard/budget_control_dashboard.py
# Copyright (c) 2026, PT. Inovasi Terbaik Bangsa and contributors
# For license information, please see license.txt

"""
Budget Control Dashboard - Consolidated Hybrid Report

HYBRID APPROACH:
This report combines data from multiple sources to provide complete budget visibility:

1. Native ERPNext Budget (allocated amount)
   - Source: Budget and Budget Account doctypes
   - Shows: Total allocated budget per Cost Center + Account

2. GL Entry (actual spent)
   - Source: Posted GL Entries
   - Shows: Actual expenses already posted to ledger

3. Budget Control Entry (reserved/locked amount)
   - Source: Budget Control Entry doctype (custom ledger)
   - Shows: Reserved amounts from approved Expense Requests (not yet in GL)

FIELD COMPATIBILITY:
Report columns are aligned with Budget Control Entry structure:
- Dimensions: fiscal_year, cost_center, account, project, branch
- Aggregates: allocated, actual, reserved, committed, available, variance
- Status: Dynamic status based on utilization

Users can drill-down from this summary to detailed Budget Control Entries
by clicking the "View Budget Control Entries" button with auto-filtered list.

This is the MAIN report that users should use - combines all budget tracking in one place.
"""

import frappe
from frappe import _
from imogi_finance.budget_control import ledger, utils


def execute(filters=None):
    validate_filters(filters)
    columns = get_columns()
    data = get_data(filters)
    chart = get_chart_data(data, filters)
    summary = get_summary(data)
    message = get_info_message()
    
    return columns, data, message, chart, summary


def validate_filters(filters):
    """Validate required filters."""
    if not filters:
        filters = {}
    
    if not filters.get("company"):
        filters["company"] = frappe.defaults.get_user_default("Company")
        if not filters["company"]:
            frappe.throw(_("Please select a Company"))
    
    if not filters.get("fiscal_year"):
        # Try to get current fiscal year
        try:
            from erpnext.accounts.utils import get_fiscal_year
            fiscal_year = get_fiscal_year(date=frappe.utils.today(), company=filters["company"])
            if fiscal_year:
                filters["fiscal_year"] = fiscal_year[0]
        except:
            pass
        
        if not filters.get("fiscal_year"):
            frappe.throw(_("Please select a Fiscal Year"))


def get_info_message():
    """Display information banner about how to read the dashboard."""
    return """
        <div style="padding: 12px; background: #e8f5e9; border-left: 4px solid #4caf50; margin-bottom: 15px; border-radius: 4px;">
            <h4 style="margin: 0 0 10px 0; color: #2e7d32;">📊 Cara Membaca Budget Control Dashboard</h4>
            <div style="font-size: 13px; line-height: 1.6; color: #1b5e20;">
                <p style="margin: 5px 0;"><strong>Formula Perhitungan:</strong></p>
                <ul style="margin: 5px 0 10px 20px;">
                    <li><strong>Net Reserved</strong> = Reservation - Consumption + Reversal</li>
                    <li><strong>Committed</strong> = Actual (GL) + Net Reserved</li>
                    <li><strong>Available</strong> = Allocated - Committed</li>
                </ul>
                <p style="margin: 5px 0;"><strong>Arti Kolom:</strong></p>
                <ul style="margin: 5px 0 10px 20px;">
                    <li><strong>Reservation (+)</strong>: Budget di-lock untuk Expense Request yang approved</li>
                    <li><strong>Consumption (-)</strong>: Budget yang dikonsumsi oleh Purchase Invoice yang submitted</li>
                    <li><strong>Reversal (+)</strong>: Budget yang dikembalikan saat Purchase Invoice di-cancel</li>
                    <li><strong>Net Reserved</strong>: Budget yang masih di-hold untuk ER yang belum jadi PI</li>
                    <li><strong>Committed</strong>: Total budget terpakai (actual + reserved)</li>
                    <li><strong>Available</strong>: Budget yang masih bisa digunakan untuk ER baru</li>
                </ul>
                <p style="margin: 5px 0;"><strong>Status Budget:</strong> 
                    <span style="color: #4caf50;">✓ In Use (<75%)</span> | 
                    <span style="color: #ff9800;">⚠ Warning (75-90%)</span> | 
                    <span style="color: #f44336;">✗ Critical (>90%)</span>
                </p>
            </div>
        </div>
    """


def get_columns():
    """Define report columns - combines native budget + Budget Control Entry breakdown."""
    return [
        {
            "fieldname": "cost_center",
            "label": _("Cost Center"),
            "fieldtype": "Link",
            "options": "Cost Center",
            "width": 200
        },
        {
            "fieldname": "account",
            "label": _("Account"),
            "fieldtype": "Link",
            "options": "Account",
            "width": 200
        },
        {
            "fieldname": "project",
            "label": _("Project"),
            "fieldtype": "Link",
            "options": "Project",
            "width": 150
        },
        {
            "fieldname": "branch",
            "label": _("Branch"),
            "fieldtype": "Link",
            "options": "Branch",
            "width": 150
        },
        {
            "fieldname": "allocated",
            "label": _("Allocated"),
            "fieldtype": "Currency",
            "width": 120,
            "description": _("Total budget yang dialokasikan dari ERPNext Budget")
        },
        {
            "fieldname": "actual",
            "label": _("Actual (GL)"),
            "fieldtype": "Currency",
            "width": 120,
            "description": _("Pengeluaran aktual yang sudah di-posting ke General Ledger (PI, Payment)")
        },
        {
            "fieldname": "reservation",
            "label": _("Reservation (+)"),
            "fieldtype": "Currency",
            "width": 120,
            "description": _("Budget yang di-lock untuk Expense Request yang approved (entry OUT)")
        },
        {
            "fieldname": "consumption",
            "label": _("Consumption (-)"),
            "fieldtype": "Currency",
            "width": 120,
            "description": _("Budget yang dikonsumsi saat Purchase Invoice submit (entry IN, mengurangi reservation)")
        },
        {
            "fieldname": "reversal",
            "label": _("Reversal (+)"),
            "fieldtype": "Currency",
            "width": 120,
            "description": _("Budget yang dikembalikan saat Purchase Invoice di-cancel (entry OUT, restore reservation)")
        },
        {
            "fieldname": "reclass",
            "label": _("Reclass"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "supplement",
            "label": _("Supplement"),
            "fieldtype": "Currency",
            "width": 120
        },
        {
            "fieldname": "net_reserved",
            "label": _("Net Reserved"),
            "fieldtype": "Currency",
            "width": 120,
            "description": _("Reservation - Consumption + Reversal = Budget masih di-hold untuk ER belum jadi PI")
        },
        {
            "fieldname": "committed",
            "label": _("Committed"),
            "fieldtype": "Currency",
            "width": 120,
            "description": _("Actual + Net Reserved = Total budget yang sudah terpakai atau akan terpakai")
        },
        {
            "fieldname": "committed_pct",
            "label": _("Committed %"),
            "fieldtype": "Percent",
            "width": 100,
            "description": _("Persentase committed dari allocated (target <75% = aman, >90% = kritis)")
        },
        {
            "fieldname": "available",
            "label": _("Available"),
            "fieldtype": "Currency",
            "width": 120,
            "description": _("Allocated - Committed = Budget yang masih bisa digunakan untuk ER baru")
        },
        {
            "fieldname": "status",
            "label": _("Status"),
            "fieldtype": "Data",
            "width": 120,
            "description": _("Status budget: Unused/In Use/Warning/Critical/Over Budget")
        }
    ]


def get_budget_control_breakdown(dims, from_date=None, to_date=None):
    """
    Get breakdown of Budget Control Entry by entry_type.
    
    Simplified Flow (RELEASE deprecated, use RESERVATION IN instead):
    - RESERVATION OUT: Budget locked for Expense Request
    - RESERVATION IN: Budget released on ER rejection/cancel (replaces RELEASE)
    - CONSUMPTION IN: Consumes reserved budget when PI submit
    - REVERSAL OUT: Restores reserved budget when PI cancel
    - RECLASS: Budget movement between accounts
    - SUPPLEMENT: Additional budget allocation
    
    Reserved = RESERVATION(OUT) - RESERVATION(IN) - CONSUMPTION(IN) + REVERSAL(OUT)
    
    Note: breakdown values are pre-signed based on direction:
    - OUT = positive (adds to breakdown)
    - IN = negative (subtracts from breakdown)
    
    So net_reserved = breakdown["reservation"] + breakdown["consumption"] + breakdown["reversal"]
    works correctly because signs are already applied.
    
    Returns dict with keys: reservation, consumption, reversal, reclass, supplement
    """
    filters = {
        "company": dims.company,
        "fiscal_year": dims.fiscal_year,
        "cost_center": dims.cost_center,
        "account": dims.account,
        "docstatus": 1
    }
    
    if dims.project:
        filters["project"] = dims.project
    if dims.branch:
        filters["branch"] = dims.branch
    if from_date and to_date:
        filters["posting_date"] = ["between", [from_date, to_date]]
    
    try:
        entries = frappe.get_all(
            "Budget Control Entry",
            filters=filters,
            fields=["entry_type", "direction", "amount"]
        )
    except:
        entries = []
    
    breakdown = {
        "reservation": 0.0,
        "consumption": 0.0,
        "reversal": 0.0,
        "reclass": 0.0,
        "supplement": 0.0
    }
    
    for entry in entries:
        entry_type = entry.get("entry_type", "").lower()
        direction = entry.get("direction")
        amount = float(entry.get("amount") or 0.0)
        
        if entry_type in breakdown:
            # Signed amount based on direction
            if direction == "OUT":
                breakdown[entry_type] += amount  # OUT = positive (reserve/restore)
            elif direction == "IN":
                breakdown[entry_type] -= amount  # IN = negative (consume)
    
    return breakdown


def get_data(filters):
    """
    Get consolidated budget data with Budget Control Entry breakdown.
    
    Combines:
    1. Budget allocation (from native Budget)
    2. Actual spending (from GL Entry)
    3. Budget Control Entry breakdown (RESERVATION, CONSUMPTION, RELEASE, etc.)
    """
    company = filters.get("company")
    fiscal_year = filters.get("fiscal_year")
    
    # Get all budgets for company/fiscal year
    budget_filters = {
        "company": company,
        "fiscal_year": fiscal_year,
        "docstatus": 1
    }
    
    if filters.get("cost_center"):
        budget_filters["cost_center"] = filters.get("cost_center")
    
    budgets = frappe.get_all(
        "Budget",
        filters=budget_filters,
        fields=["name", "cost_center"]
    )
    
    if not budgets:
        frappe.msgprint(_("No Budget found for {0} - {1}").format(company, fiscal_year))
        return []
    
    data = []
    from_date = filters.get("from_date")
    to_date = filters.get("to_date")
    
    for budget in budgets:
        # Get budget accounts from native Budget
        budget_accounts = frappe.get_all(
            "Budget Account",
            filters={"parent": budget.name},
            fields=["account", "budget_amount"]
        )
        
        for ba in budget_accounts:
            # Apply filters
            if filters.get("account") and ba.account != filters.get("account"):
                continue
            
            # Build dimensions
            dims = utils.Dimensions(
                company=company,
                fiscal_year=fiscal_year,
                cost_center=budget.cost_center,
                account=ba.account,
                project=filters.get("project"),
                branch=filters.get("branch")
            )
            
            # Get availability data (combines all 3 sources)
            availability = ledger.get_availability(dims, from_date=from_date, to_date=to_date)
            
            allocated = availability["allocated"]
            actual = availability["actual"]
            reserved = availability["reserved"]
            available = availability["available"]
            
            # Get Budget Control Entry breakdown by entry_type
            breakdown = get_budget_control_breakdown(dims, from_date=from_date, to_date=to_date)
            
            # Net reserved = RESERVATION(OUT-IN) - CONSUMPTION + REVERSAL
            # Simplified flow: RESERVATION IN replaces RELEASE
            # - RESERVATION OUT: +100 (locked for ER)
            # - RESERVATION IN: -100 (released on ER reject/cancel)
            # - CONSUMPTION IN: -100 (consumed by PI)
            # - REVERSAL OUT: +100 (restored on PI cancel)
            # Note: breakdown values are pre-signed (OUT=+, IN=-)
            net_reserved = breakdown["reservation"] + breakdown["consumption"] + breakdown["reversal"]
            
            # Calculate committed = actual + net reserved
            committed = actual + net_reserved
            committed_pct = (committed / allocated * 100) if allocated > 0 else 0
            
            # Skip if no activity (optional filter)
            if filters.get("hide_zero") and allocated == 0 and actual == 0 and net_reserved == 0:
                continue
            
            # Determine status
            status = get_status(allocated, actual, net_reserved, available)
            
            data.append({
                "cost_center": budget.cost_center,
                "account": ba.account,
                "project": dims.project or "",
                "branch": dims.branch or "",
                "allocated": allocated or 0,
                "actual": actual or 0,
                "reservation": breakdown["reservation"] or 0,
                "consumption": breakdown["consumption"] or 0,
                "reversal": breakdown["reversal"] or 0,
                "reclass": breakdown["reclass"] or 0,
                "supplement": breakdown["supplement"] or 0,
                "net_reserved": net_reserved or 0,
                "committed": committed or 0,
                "committed_pct": committed_pct or 0,
                "available": available or 0,
                "status": status
            })
    
    # Sort by allocated descending, then by committed percentage
    data.sort(key=lambda x: (x["allocated"], x["committed_pct"]), reverse=True)
    
    return data


def get_status(allocated, actual, reserved, available):
    """Determine budget status."""
    if allocated <= 0:
        return "No Budget"
    
    committed_pct = ((actual + reserved) / allocated) * 100
    
    if available < 0:
        return "Over Budget"
    elif available == 0:
        return "Fully Used"
    elif committed_pct >= 100:
        return "Over Committed"
    elif committed_pct > 90:
        return "Critical"
    elif committed_pct > 75:
        return "Warning"
    elif actual > 0 or reserved > 0:
        return "In Use"
    else:
        return "Unused"


def get_chart_data(data, filters):
    """Generate stacked bar chart for top 10 accounts with Budget Control Entry breakdown."""
    if not data:
        return None
    
    # Take top 10 by committed amount
    top_data = sorted(data, key=lambda x: x.get("committed", 0) or 0, reverse=True)[:10]
    
    if not top_data:
        return None
    
    labels = []
    actual_values = []
    reservation_values = []
    consumption_values = []
    available_values = []
    
    for row in top_data:
        # Shorten labels for better display
        cc = row.get('cost_center', '') or ''
        acc = row.get('account', '') or ''
        cc_short = cc.split(' - ')[-1][:12] if cc else "N/A"
        acc_short = acc.split(' - ')[-1][:15] if acc else "N/A"
        label = f"{cc_short} - {acc_short}"
        
        labels.append(label)
        actual_values.append(float(row.get("actual", 0) or 0))
        # Show absolute values for chart
        reservation_values.append(abs(float(row.get("reservation", 0) or 0)))
        consumption_values.append(abs(float(row.get("consumption", 0) or 0)))
        available_values.append(max(0, float(row.get("available", 0) or 0)))  # Don't show negative in chart
    
    chart_data = {
        "data": {
            "labels": labels,
            "datasets": [
                {
                    "name": _("Actual (GL)"),
                    "values": actual_values
                },
                {
                    "name": _("Reservation"),
                    "values": reservation_values
                },
                {
                    "name": _("Consumption"),
                    "values": consumption_values
                },
                {
                    "name": _("Available"),
                    "values": available_values
                }
            ]
        },
        "type": "bar",
        "barOptions": {
            "stacked": 1,
            "spaceRatio": 0.5
        },
        "height": 350,
        "colors": ["#FF6B6B", "#FFA726", "#42A5F5", "#66BB6A"]
    }
    
    return chart_data


def get_summary(data):
    """Generate summary cards with Budget Control Entry breakdown.
    
    Note: With simplified flow, RESERVATION includes both lock (OUT) and release (IN).
    The 'reservation' field in data is already net (OUT - IN).
    'release' field is deprecated (always 0 for new entries).
    """
    if not data:
        return []
    
    total_allocated = sum(row.get("allocated", 0) or 0 for row in data)
    total_actual = sum(row.get("actual", 0) or 0 for row in data)
    total_reservation = sum(row.get("reservation", 0) or 0 for row in data)
    total_consumption = sum(row.get("consumption", 0) or 0 for row in data)
    # RELEASE is deprecated - kept for backward compatibility with old data
    total_release = sum(row.get("release", 0) or 0 for row in data)
    total_net_reserved = sum(row.get("net_reserved", 0) or 0 for row in data)
    total_committed = sum(row.get("committed", 0) or 0 for row in data)
    total_available = sum(row.get("available", 0) or 0 for row in data)
    
    over_budget_count = len([r for r in data if (r.get("available", 0) or 0) < 0])
    critical_count = len([r for r in data if r.get("status") in ("Critical", "Warning")])
    
    summary_cards = [
        {
            "value": total_allocated,
            "indicator": "blue",
            "label": _("Total Allocated"),
            "datatype": "Currency"
        },
        {
            "value": total_actual,
            "indicator": "orange",
            "label": _("Total Actual (GL)"),
            "datatype": "Currency"
        },
        {
            "value": abs(total_reservation),
            "indicator": "red",
            "label": _("Total Reservation (Net)"),  # Includes both lock and release
            "datatype": "Currency"
        },
        {
            "value": abs(total_consumption),
            "indicator": "blue",
            "label": _("Total Consumption"),
            "datatype": "Currency"
        },
        {
            "value": abs(total_net_reserved),
            "indicator": "purple",
            "label": _("Net Reserved"),
            "datatype": "Currency"
        },
        {
            "value": total_committed,
            "indicator": "orange",
            "label": _("Total Committed"),
            "datatype": "Currency"
        },
        {
            "value": total_available,
            "indicator": "green" if total_available > 0 else "red",
            "label": _("Total Available"),
            "datatype": "Currency"
        },
        {
            "value": over_budget_count,
            "indicator": "red" if over_budget_count > 0 else "grey",
            "label": _("Over Budget Accounts")
        },
        {
            "value": critical_count,
            "indicator": "yellow" if critical_count > 0 else "grey",
            "label": _("Critical/Warning Accounts")
        }
    ]
    
    # Add legacy Release card only if there's data (backward compatibility)
    if total_release != 0:
        summary_cards.insert(4, {
            "value": abs(total_release),
            "indicator": "green",
            "label": _("Total Release (Legacy)"),
            "datatype": "Currency"
        })
    
    return summary_cards
